### ISSUE TITLE
Fix Duplication

### ISSUE THREAD

**Issue by RichLogan** [RichLogan/CiscoSpark-UnitySDK#10]
There is a tonne of duplication that should be removed maybe via a `SparkObject` parent giving implementations of:
- Commit
- Delete
- List
- GetDetails

as they are all the same except resource name and fields.

**Comment by RichLogan** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Inheritance from SparkObject to stop all the duplication](https://trello.com/c/7w058z09/11-inheritance-from-sparkobject-to-stop-all-the-duplication)

**Comment by RichLogan** [comment]
I am actually working on this, using a base `SparkObject` class that uses generics to return the different Spark types. Running into some issues with Static properties with generics and other similar things, so want to make sure I do it the correct way.

I'll add this in with backward compatibility initially but will deprecate the old specific calls such as `Membership.GetMembershipDetails` to just become `AnySparkObjectChild.GetDetails` when v2.0 comes.

**Comment by RichLogan** [comment]
Making an effort to smash this out this week, made a good start today on the [`oop-rewrite`](https://github.com/RichLogan/CiscoSpark-UnitySDK/tree/oop-rewrite) branch. I hit this huge roadblock because I thought Generics were the answer to all my problems 😛 

TL;DR - `Request` now handles all web requests, `SparkObject` parent for the Spark classes provides most default functionality and fields and the individual classes such as `Room` cover anything specific to themselves. 

Couple of other helper classes that handle things like API endpoints, which fields can actually be updated etc. 

API will pretty much change completely to be much more OOP style.

**Comment by RichLogan** [comment]
Since oop-rewrite has been merged this is closing. Any related issues can be created under a more specific issue/branch/PR
